File: server/src/services/statistics.service.ts
import { Attendance, User, Schedule, Workshift } from "@models";
import { AttendanceStatus } from "@models/attendance";
import { Op } from "sequelize";
import { format, startOfMonth, endOfMonth, differenceInMinutes } from "date-fns";

interface PersonalStats {
	checkInTime: string | null;
	checkOutTime: string | null;
	totalTime: string;
	isCheckedIn: boolean;
	daysWorked: number;
	daysOff: number;
	overtimeHours: string;
	lateArrivals: number;
}

interface TeamStats {
	teamPresent: number;
	teamLate: number;
	teamAbsent: number;
	total: number;
}

interface TodayShift {
	checkInTime: string | null;
	checkOutTime: string | null;
	totalTime: string;
	isCheckedIn: boolean;
	status: AttendanceStatus | null;
}

export default class StatisticsService {
	/**
	 * Get current shift status for today
	 */
	static async getTodayShift(userId: string): Promise<TodayShift> {
		const today = format(new Date(), "yyyy-MM-dd");
		
		const attendance = await Attendance.findOne({
			where: {
				userId,
				date: today,
			},
		});

		if (!attendance) {
			return {
				checkInTime: null,
				checkOutTime: null,
				totalTime: "--:--",
				isCheckedIn: false,
				status: null,
			};
		}

		const checkInTime = attendance.checkInTime
			? format(attendance.checkInTime, "HH:mm")
			: null;
		const checkOutTime = attendance.checkOutTime
			? format(attendance.checkOutTime, "HH:mm")
			: null;
		const isCheckedIn = !!attendance.checkInTime && !attendance.checkOutTime;

		let totalTime = "--:--";
		if (attendance.checkInTime) {
			const endTime = attendance.checkOutTime || new Date();
			const minutes = differenceInMinutes(endTime, attendance.checkInTime);
			const hours = Math.floor(minutes / 60);
			const mins = minutes % 60;
			totalTime = `${hours}h ${mins}m`;
		}

		return {
			checkInTime,
			checkOutTime,
			totalTime,
			isCheckedIn,
			status: attendance.status,
		};
	}

	/**
	 * Get personal monthly statistics
	 */
	static async getPersonalStats(userId: string, month?: number, year?: number): Promise<PersonalStats> {
		const now = new Date();
		const targetMonth = month || now.getMonth() + 1;
		const targetYear = year || now.getFullYear();

		const startDate = startOfMonth(new Date(targetYear, targetMonth - 1, 1));
		const endDate = endOfMonth(startDate);

		// Get today's shift
		const todayShift = await this.getTodayShift(userId);

		// Get all attendance records for the month
		const attendances = await Attendance.findAll({
			where: {
				userId,
				date: {
					[Op.between]: [format(startDate, "yyyy-MM-dd"), format(endDate, "yyyy-MM-dd")],
				},
			},
			include: [
				{
					model: Schedule,
					as: "schedule",
					include: [
						{
							model: Workshift,
							as: "Shift",
						},
					],
				},
			],
		});

		// Calculate statistics
		let daysWorked = 0;
		let daysOff = 0;
		let lateArrivals = 0;
		let totalOvertimeMinutes = 0;

		attendances.forEach((att) => {
			if (att.status === AttendanceStatus.PRESENT || att.status === AttendanceStatus.LATE) {
				daysWorked++;
			}
			if (att.status === AttendanceStatus.LATE) {
				lateArrivals++;
			}
			if (att.status === AttendanceStatus.ABSENT) {
				daysOff++;
			}

			// Calculate overtime
			if (att.checkInTime && att.checkOutTime && att.schedule?.Shift) {
				const shift = att.schedule.Shift;
				const checkOutTime = att.checkOutTime;
				
				// Parse shift end time (assuming it's a TIME field)
				const shiftEndStr = shift.endTime.toString();
				const timeParts = shiftEndStr.split(':').map(Number);
				const endHour = timeParts[0];
				const endMinute = timeParts[1];
				
				if (endHour !== undefined && endMinute !== undefined) {
					const shiftEndDate = new Date(att.date);
					shiftEndDate.setHours(endHour, endMinute, 0, 0);

					// If checked out after shift end time, count as overtime
					if (checkOutTime > shiftEndDate) {
						const overtimeMinutes = differenceInMinutes(checkOutTime, shiftEndDate);
						if (overtimeMinutes > 0) {
							totalOvertimeMinutes += overtimeMinutes;
						}
					}
				}
			}
		});

		const overtimeHours = totalOvertimeMinutes > 0
			? `${Math.floor(totalOvertimeMinutes / 60)}h`
			: "0h";

		return {
			checkInTime: todayShift.checkInTime,
			checkOutTime: todayShift.checkOutTime,
			totalTime: todayShift.totalTime,
			isCheckedIn: todayShift.isCheckedIn,
			daysWorked,
			daysOff,
			overtimeHours,
			lateArrivals,
		};
	}

	/**
	 * Get team statistics for today (for managers/admins)
	 */
	static async getTeamStats(): Promise<TeamStats> {
		const today = format(new Date(), "yyyy-MM-dd");

		// Get all active users
		const activeUsers = await User.count({
			where: {
				status: "ACTIVE",
			},
		});

		// Get today's attendance records
		const attendances = await Attendance.findAll({
			where: {
				date: today,
			},
		});

		// Count by status
		let teamPresent = 0;
		let teamLate = 0;
		let teamAbsent = 0;

		attendances.forEach((att) => {
			if (att.status === AttendanceStatus.PRESENT) {
				teamPresent++;
			} else if (att.status === AttendanceStatus.LATE) {
				teamLate++;
			} else if (att.status === AttendanceStatus.ABSENT) {
				teamAbsent++;
			}
		});

		return {
			teamPresent,
			teamLate,
			teamAbsent,
			total: activeUsers,
		};
	}

	/**
	 * Get detailed team attendance for today (for realtime dashboard)
	 */
	static async getTeamAttendanceDetails() {
		const today = format(new Date(), "yyyy-MM-dd");

		const attendances = await Attendance.findAll({
			where: {
				date: today,
			},
			include: [
				{
					model: User,
					as: "user",
					attributes: ["id", "name", "email", "position", "department"],
				},
			],
			order: [["checkInTime", "DESC"]],
		});

		return attendances.map((att) => ({
			userId: att.userId,
			userName: att.user?.name || "Unknown",
			userEmail: att.user?.email,
			position: att.user?.position,
			department: att.user?.department,
			checkInTime: att.checkInTime ? format(att.checkInTime, "HH:mm") : null,
			checkOutTime: att.checkOutTime ? format(att.checkOutTime, "HH:mm") : null,
			status: att.status,
			isCheckedIn: !!att.checkInTime && !att.checkOutTime,
		}));
	}
}

File: client/lib/src/services/statistics.service.dart
import 'package:dio/dio.dart';
import '../consts/api_endpoints.dart';
import '../utils/api_client.dart';

class PersonalStats {
  final String? checkInTime;
  final String? checkOutTime;
  final String totalTime;
  final bool isCheckedIn;
  final int daysWorked;
  final int daysOff;
  final String overtimeHours;
  final int lateArrivals;

  PersonalStats({
    this.checkInTime,
    this.checkOutTime,
    required this.totalTime,
    required this.isCheckedIn,
    required this.daysWorked,
    required this.daysOff,
    required this.overtimeHours,
    required this.lateArrivals,
  });

  factory PersonalStats.fromJson(Map<String, dynamic> json) {
    return PersonalStats(
      checkInTime: json['checkInTime'],
      checkOutTime: json['checkOutTime'],
      totalTime: json['totalTime'] ?? '--:--',
      isCheckedIn: json['isCheckedIn'] ?? false,
      daysWorked: json['daysWorked'] ?? 0,
      daysOff: json['daysOff'] ?? 0,
      overtimeHours: json['overtimeHours'] ?? '0h',
      lateArrivals: json['lateArrivals'] ?? 0,
    );
  }
}

class TodayShift {
  final String? checkInTime;
  final String? checkOutTime;
  final String totalTime;
  final bool isCheckedIn;
  final String? status;

  TodayShift({
    this.checkInTime,
    this.checkOutTime,
    required this.totalTime,
    required this.isCheckedIn,
    this.status,
  });

  factory TodayShift.fromJson(Map<String, dynamic> json) {
    return TodayShift(
      checkInTime: json['checkInTime'],
      checkOutTime: json['checkOutTime'],
      totalTime: json['totalTime'] ?? '--:--',
      isCheckedIn: json['isCheckedIn'] ?? false,
      status: json['status'],
    );
  }
}

class TeamStats {
  final int teamPresent;
  final int teamLate;
  final int teamAbsent;
  final int total;

  TeamStats({
    required this.teamPresent,
    required this.teamLate,
    required this.teamAbsent,
    required this.total,
  });

  factory TeamStats.fromJson(Map<String, dynamic> json) {
    return TeamStats(
      teamPresent: json['teamPresent'] ?? 0,
      teamLate: json['teamLate'] ?? 0,
      teamAbsent: json['teamAbsent'] ?? 0,
      total: json['total'] ?? 0,
    );
  }
}

class TeamAttendanceDetail {
  final String userId;
  final String userName;
  final String? userEmail;
  final String? position;
  final String? department;
  final String? checkInTime;
  final String? checkOutTime;
  final String status;
  final bool isCheckedIn;

  TeamAttendanceDetail({
    required this.userId,
    required this.userName,
    this.userEmail,
    this.position,
    this.department,
    this.checkInTime,
    this.checkOutTime,
    required this.status,
    required this.isCheckedIn,
  });

  factory TeamAttendanceDetail.fromJson(Map<String, dynamic> json) {
    return TeamAttendanceDetail(
      userId: json['userId'],
      userName: json['userName'] ?? 'Unknown',
      userEmail: json['userEmail'],
      position: json['position'],
      department: json['department'],
      checkInTime: json['checkInTime'],
      checkOutTime: json['checkOutTime'],
      status: json['status'] ?? 'ABSENT',
      isCheckedIn: json['isCheckedIn'] ?? false,
    );
  }
}

class StatisticsService {
  StatisticsService({Dio? dio})
      : _dio = dio ?? ApiClient().client;

  final Dio _dio;

  /// Get personal statistics for a user
  Future<PersonalStats> getPersonalStats(
    String userId, {
    int? month,
    int? year,
  }) async {
    final queryParams = <String, dynamic>{};
    if (month != null) queryParams['month'] = month;
    if (year != null) queryParams['year'] = year;

    final response = await _dio.get(
      '${ApiEndpoints.statistics}/personal/$userId',
      queryParameters: queryParams,
    );

    if (response.data['success'] == true) {
      return PersonalStats.fromJson(response.data['data']);
    }

    throw Exception('Failed to fetch personal statistics');
  }

  /// Get today's shift status for a user
  Future<TodayShift> getTodayShift(String userId) async {
    final response = await _dio.get(
      '${ApiEndpoints.statistics}/today/$userId',
    );

    if (response.data['success'] == true) {
      return TodayShift.fromJson(response.data['data']);
    }

    throw Exception('Failed to fetch today\'s shift');
  }

  /// Get team statistics (managers/admins only)
  Future<TeamStats> getTeamStats() async {
    final response = await _dio.get(
      '${ApiEndpoints.statistics}/team',
    );

    if (response.data['success'] == true) {
      return TeamStats.fromJson(response.data['data']);
    }

    throw Exception('Failed to fetch team statistics');
  }

  /// Get detailed team attendance (managers/admins only)
  Future<List<TeamAttendanceDetail>> getTeamAttendanceDetails() async {
    final response = await _dio.get(
      '${ApiEndpoints.statistics}/team/details',
    );

    if (response.data['success'] == true) {
      final data = response.data['data'] as List;
      return data
          .map((json) => TeamAttendanceDetail.fromJson(json))
          .toList();
    }

    throw Exception('Failed to fetch team attendance details');
  }
}
